From: Fabian Pflug <f.pflug@pengutronix.de>
Date: Wed, 15 Apr 2026 09:36:44 +0200
Subject: [PATCH] core: rpmb_fs: make tee_rpmb_key_gen function public

Remove the static from tee_rpmb_key_gen and add it to the header in
order to make it available for other parts of the code to generate the
RPMB key based on a provided CID.

Signed-off-by: Fabian Pflug <f.pflug@pengutronix.de>
---
 core/include/tee/tee_fs.h |  8 ++++++++
 core/tee/tee_rpmb_fs.c    | 26 +++++++++++---------------
 2 files changed, 19 insertions(+), 15 deletions(-)

diff --git a/core/include/tee/tee_fs.h b/core/include/tee/tee_fs.h
index 09b172794efe..36d012b9c85d 100644
--- a/core/include/tee/tee_fs.h
+++ b/core/include/tee/tee_fs.h
@@ -67,6 +67,14 @@ TEE_Result tee_rpmb_reinit(void);
 /* Ger RPMB memory allocation statistics */
 TEE_Result rpmb_mem_stats(struct pta_stats_alloc *stats, bool reset);
 
+#define RPMB_EMMC_CID_SIZE 16
+
+/*
+ * Generate the RPMB key for the given cid. The cid must be RPMB_EMMC_CID_SIZE
+ * bytes long.
+ */
+TEE_Result tee_rpmb_key_gen(uint8_t *cid, uint8_t *key, uint32_t len);
+
 /**
  * Weak function which can be overridden by platforms to indicate that the RPMB
  * key is ready to be written. Defaults to true, platforms can return false to
diff --git a/core/tee/tee_rpmb_fs.c b/core/tee/tee_rpmb_fs.c
index f66f79b935b3..4ea274f71c6e 100644
--- a/core/tee/tee_rpmb_fs.c
+++ b/core/tee/tee_rpmb_fs.c
@@ -232,7 +232,6 @@ struct rpmb_raw_data {
 	uint8_t byte_offset;
 };
 
-#define RPMB_EMMC_CID_SIZE 16
 struct rpmb_dev_info {
 	uint8_t cid[RPMB_EMMC_CID_SIZE];
 	/* EXT CSD-slice 168 "RPMB Size" */
@@ -297,29 +296,24 @@ static const uint8_t rpmb_test_key[RPMB_KEY_MAC_SIZE] = {
 	0xAB, 0x56, 0xE6, 0xC6, 0x1B, 0xB7, 0x01, 0xE4
 };
 
-static TEE_Result tee_rpmb_key_gen(uint8_t *key, uint32_t len)
+TEE_Result tee_rpmb_key_gen(uint8_t *cid __unused, uint8_t *key, uint32_t len)
 {
-	TEE_Result res = TEE_SUCCESS;
-
-	if (!key || RPMB_KEY_MAC_SIZE != len) {
-		res = TEE_ERROR_BAD_PARAMETERS;
-		goto out;
-	}
+	if (!key || RPMB_KEY_MAC_SIZE != len)
+		return TEE_ERROR_BAD_PARAMETERS;
 
 	DMSG("RPMB: Using test key");
 	memcpy(key, rpmb_test_key, RPMB_KEY_MAC_SIZE);
 
-out:
-	return res;
+	return TEE_SUCCESS;
 }
 
 #else /* !CFG_RPMB_TESTKEY */
 
-static TEE_Result tee_rpmb_key_gen(uint8_t *key, uint32_t len)
+TEE_Result tee_rpmb_key_gen(uint8_t *cid, uint8_t *key, uint32_t len)
 {
 	uint8_t message[RPMB_EMMC_CID_SIZE];
 
-	if (!key || RPMB_KEY_MAC_SIZE != len)
+	if (!cid || !key || RPMB_KEY_MAC_SIZE != len)
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	IMSG("RPMB: Using generated key");
@@ -332,7 +326,7 @@ static TEE_Result tee_rpmb_key_gen(uint8_t *key, uint32_t len)
 	 * CID [07: 01]: CRC (CRC7 checksum)
 	 * CID [00]: not used
 	 */
-	memcpy(message, rpmb_ctx->cid, RPMB_EMMC_CID_SIZE);
+	memcpy(message, cid, RPMB_EMMC_CID_SIZE);
 	memset(message + RPMB_CID_PRV_OFFSET, 0, 1);
 	memset(message + RPMB_CID_CRC_OFFSET, 0, 1);
 	return huk_subkey_derive(HUK_SUBKEY_RPMB, message, sizeof(message),
@@ -1162,7 +1156,8 @@ static TEE_Result legacy_rpmb_init(void)
 	if (!rpmb_ctx->key_derived) {
 		DMSG("RPMB INIT: Deriving key");
 
-		res = tee_rpmb_key_gen(rpmb_ctx->key, RPMB_KEY_MAC_SIZE);
+		res = tee_rpmb_key_gen(rpmb_ctx->cid, rpmb_ctx->key,
+				       RPMB_KEY_MAC_SIZE);
 		if (res != TEE_SUCCESS) {
 			EMSG("RPMB INIT: Deriving key failed with error 0x%x",
 				res);
@@ -1268,7 +1263,8 @@ next:
 			continue;
 		}
 
-		res = tee_rpmb_key_gen(rpmb_ctx->key, RPMB_KEY_MAC_SIZE);
+		res = tee_rpmb_key_gen(rpmb_ctx->cid, rpmb_ctx->key,
+				       RPMB_KEY_MAC_SIZE);
 		if (res)
 			return res;
 
